// @smartthings/core-sdk/dist/endpoint/modes.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModesEndpoint = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
const types_1 = require("../types");
class ModesEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('locations', config));
    }
    /**
     * Returns a list of the modes defined for a location
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    list(locationId) {
        return this.client.getPagedItems(`${this.locationId(locationId)}/modes`);
    }
    /**
     * Returns a specific mode
     * @param id UUID of the mode
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    get(id, locationId) {
        return __awaiter(this, void 0, void 0, function* () {
            const list = yield this.list(locationId);
            if (list) {
                const item = list.find(it => it.id === id);
                if (item) {
                    return item;
                }
            }
            throw Error(`Mode ${id} not found`);
        });
    }
    /**
     * Returns the currently active mode of a location
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    getCurrent(locationId) {
        return this.client.get(`${this.locationId(locationId)}/modes/current`);
    }
    /**
     * Sets the currently active mode of a location
     * @param id UUID of the mode
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    setCurrent(id, locationId) {
        return this.client.put(`${this.locationId(locationId)}/modes/current`, { modeId: id });
    }
    /**
     * Create a new mode in a location
     * @param data definition specifying the name of the new mode
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    create(data, locationId) {
        return this.client.post(`${this.locationId(locationId)}/modes`, data);
    }
    /**
     * Updates the name of a mode
     * @param id UUID of the mode
     * @param data definition specifying the new mode name
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    update(id, data, locationId) {
        return this.client.put(`${this.locationId(locationId)}/modes/${id}`, data);
    }
    /**
     * Delete a mode
     * @param id UUID of the mode
     * @param locationId UUID of the location. If the client is configured with a locationId this parameter is
     * not necessary.
     */
    delete(id, locationId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.delete(`${this.locationId(locationId)}/modes/${id}`);
            return types_1.SuccessStatusValue;
        });
    }
}
exports.ModesEndpoint = ModesEndpoint;
