// @smartthings/core-sdk/dist/endpoint/presentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PresentationEndpoint = exports.PatchItemOpEnum = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
var PatchItemOpEnum;
(function (PatchItemOpEnum) {
    PatchItemOpEnum["ADD"] = "add";
    PatchItemOpEnum["REPLACE"] = "replace";
    PatchItemOpEnum["REMOVE"] = "remove";
})(PatchItemOpEnum = exports.PatchItemOpEnum || (exports.PatchItemOpEnum = {}));
class PresentationEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('presentation', config));
    }
    /**
     * Get or generate a device configuration based on profile.
     *
     * @param extraParams deprecated.
     */
    generate(profileId, extraParams) {
        return this.client.get(`types/${profileId}/deviceconfig`, extraParams);
    }
    /**
     * Get a device configuration
     * @param presentationId The id returned from the device config create operation
     * @param manufacturerName The manufacturer name, e.g. SmartThingsCommunity
     */
    get(presentationId, manufacturerName) {
        if (manufacturerName) {
            return this.client.get('deviceconfig', { presentationId, manufacturerName });
        }
        else {
            return this.client.get('deviceconfig', { presentationId });
        }
    }
    /**
     * Make an idempotent call to either create or update a device configuration
     * based on the structure of the provided payload.
     */
    create(deviceConfig) {
        return this.client.post('deviceconfig', deviceConfig);
    }
    /**
     * Get a device presentation. If manufacturerName is omitted the default SmartThingsCommunity manufacturerName is used.
     * @param presentationId The id returned from the device config create operation
     * @param manufacturerName The manufacturer name, e.g. SmartThingsCommunity
     */
    getPresentation(presentationId, manufacturerName) {
        if (manufacturerName) {
            return this.client.get('', { presentationId, manufacturerName });
        }
        else {
            return this.client.get('', { presentationId });
        }
    }
}
exports.PresentationEndpoint = PresentationEndpoint;
