// @smartthings/core-sdk/dist/endpoint/rooms.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomsEndpoint = exports.Room = exports.RoomRequest = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
const types_1 = require("../types");
class RoomRequest {
}
exports.RoomRequest = RoomRequest;
class Room extends RoomRequest {
}
exports.Room = Room;
class RoomsEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('locations', config));
    }
    /**
     * List the rooms in a location
     * @param locationId UUID of the location
     */
    list(locationId) {
        return this.client.getPagedItems(`${this.locationId(locationId)}/rooms`);
    }
    /**
     * Get a specific room in a location
     * @param id UUID of the room
     * @param locationId UUID of the location. If the client is configured with a location ID this parameter
     * can be omitted
     */
    get(id, locationId) {
        return this.client.get(`${this.locationId(locationId)}/rooms/${id}`);
    }
    /**
     * Create a room in a location
     * @param data request containing the room name
     * @param locationId  UUID of the location. If the client is configured with a location ID this parameter
     * can be omitted
     */
    create(data, locationId) {
        return this.client.post(`${this.locationId(locationId)}/rooms`, data);
    }
    /**
     * Update a room
     * @param id UUID of the room
     * @param data request containing the name of the room
     * @param locationId UUID of the location. If the client is configured with a location ID this parameter
     * can be omitted
     */
    update(id, data, locationId) {
        return this.client.put(`${this.locationId(locationId)}/rooms/${id}`, data);
    }
    /**
     * Delete a room from a location
     * @param id UUID of the room
     * @param locationId UUID of the location. If the client is configured with a location ID this parameter
     * can be omitted
     */
    delete(id, locationId) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.client.delete(`${this.locationId(locationId)}/rooms/${id}`);
            return types_1.SuccessStatusValue;
        });
    }
    /**
     * Returns a list of all the devices in a room
     * @param id UUID of the room
     * @param locationId UUID of the location. If the client is configured with a location ID this parameter
     * can be omitted
     */
    listDevices(id, locationId) {
        return this.client.getPagedItems(`${this.locationId(locationId)}/rooms/${id}/devices`);
    }
}
exports.RoomsEndpoint = RoomsEndpoint;
